### PR TITLE
rafthttp: deprecate streamTypeMsgApp and remove msgApp stream sent restriction due to streamTypeMsgApp

### PR DESCRIPTION
**Pull request by yichengq**
streamTypeMsgApp is only used in etcd 2.0. etcd 2.3 should not talk to
etcd 2.0, either send or receive requests. So I deprecate streamTypeMsgApp
and its related stuffs from rafthttp package.

updating term is only used from streamTypeMsgApp, so it is removed too.

Moreover, we send all MsgApp on stream msgAppV2 now. For stream msgAppV2, as long as the message is MsgApp type, it should be sent through stream msgAppV2.

### CODE REVIEW

File: rafthttp/stream.go [hunk 1]
```diff
@@ -47,9 +45,9 @@ var (
 
 	// the key is in string format "major.minor.patch"
 	supportedStream = map[string][]streamType{
-		"2.0.0": {streamTypeMsgApp},
-		"2.1.0": {streamTypeMsgApp, streamTypeMsgAppV2, streamTypeMessage},
-		"2.2.0": {streamTypeMsgApp, streamTypeMsgAppV2, streamTypeMessage},
+		"2.0.0": {},
```
Comment (xiang90) [inline rafthttp/stream.go]: can we just remove this line?

File: rafthttp/peer.go [hunk 2]
```diff
@@ -271,12 +264,14 @@ func (p *peer) pick(m raftpb.Message) (writec chan<- raftpb.Message, picked stri
 	// stream for a long time, only use one of the N pipelines to send MsgSnap.
 	if isMsgSnap(m) {
 		return p.pipeline.msgc, pipelineMsg
-	} else if writec, ok = p.msgAppWriter.writec(); ok && canUseMsgAppStream(m) {
```
Comment (xiang90) [inline rafthttp/peer.go]: i am not sure about this change. If we only change this for deprecation, shouldn't we just remove line274-275?

File: rafthttp/stream.go (hunk 1 above)
Comment (yichengq) [inline rafthttp/stream.go]: I wants to remove this line and add new `"2.3.0"` line in another commit that bumps support map to 2.3 to make it clear.

File: rafthttp/peer.go (hunk 2 above)
Comment (yichengq) [inline rafthttp/peer.go]: `msgAppWriter` is used for msgAppV2 stream after change. It deserves a better name `msgAppV2Writer` to make it clear.

Originally `msgAppWriter` is used for both msgApp and msgAppV2 stream.

Comment (xiang90) [inline rafthttp/peer.go]: No. I mean before this change, the func will return `streamApp`. After this change, this will return `streamAppV2` instead. Do you mean that the previous picked name will be incorrect when the writec is actually v2?

Comment (yichengq) [inline rafthttp/peer.go]: > Do you mean that the previous picked name will be incorrect when the writec is actually v2?

Yes. msgAppWriter may build either `streamAppV2` or `streamApp` to send append messages, and it logs as `streamApp` always.

Comment (xiang90) [inline rafthttp/peer.go]: OK. Can you please separate this into another commit? This is actually a bug fixing. Or you might think it does not worth the effort. But we need to make sure this is test-covered.

Comment (yichengq) [inline rafthttp/peer.go]: sure

**Comment by yichengq** [comment]
@xiang90 Separate the bug fix at https://github.com/yichengq/etcd/commit/33231fccdd03337a56e03e87cc4a3a7e561e6cb4 It could be implemented quite clean after deprecating streamTypeMsgApp, so i put it after deprecate commit.

It is a display bug in 2.2.1 too. It may need some extra changes to make it fixed in 2.2.1 because stream doesn't expose its stream type. My thought is that fixing it at 2.3 is good enough.

**Comment by xiang90** [comment]
@yichengq have you checked master can work well with etcd 2.2 after this change?

**Comment by yichengq** [comment]
@xiang90 Good point. Just checked that by running etcd 2.2 mixed with master. Either it is leader or follower, it could work pretty well with amounts of client requests.

**Comment by xiang90** [comment]
LGTM

**Comment by yichengq** [comment]
For semaphore failure, I am discussing with jon about it. This is not related to this issue, so I will merge it.
